conformance/utils/suite/experimental_suite.go

```diff
@@ -0,0 +1,278 @@
+		GatewayClassName: s.GatewayClassName,
+		Debug:            s.Debug,
+		Cleanup:          s.CleanupBaseResources,
+		BaseManifests:    s.BaseManifests,
+		Applier: kubernetes.Applier{
+			NamespaceLabels:          s.NamespaceLabels,
+			ValidUniqueListenerPorts: s.ValidUniqueListenerPorts,
+		},
+		SupportedFeatures: s.SupportedFeatures,
+		TimeoutConfig:     s.TimeoutConfig,
+		SkipTests:         sets.New(s.SkipTests...),
+		FS:                *s.FS,
+	}
+
+	// apply defaults
+	if suite.BaseManifests == "" {
+		suite.BaseManifests = "base/manifests.yaml"
+	}
+
+	return suite, nil
+}
+
+// -----------------------------------------------------------------------------
+// Conformance Test Suite - Public Methods
+// -----------------------------------------------------------------------------
+
+// Setup ensures the base resources required for conformance tests are installed
+// in the cluster. It also ensures that all relevant resources are ready.
+func (suite *ExperimentalConformanceTestSuite) Setup(t *testing.T) {
+	t.Logf("Test Setup: Ensuring GatewayClass has been accepted")
+	suite.ControllerName = kubernetes.GWCMustHaveAcceptedConditionTrue(t, suite.Client, suite.TimeoutConfig, suite.GatewayClassName)
+
+	suite.Applier.GatewayClass = suite.GatewayClassName
+	suite.Applier.ControllerName = suite.ControllerName
+	suite.Applier.FS = suite.FS
+
+	t.Logf("Test Setup: Applying base manifests")
+	suite.Applier.MustApplyWithCleanup(t, suite.Client, suite.TimeoutConfig, suite.BaseManifests, suite.Cleanup)
+
+	t.Logf("Test Setup: Applying programmatic resources")
+	secret := kubernetes.MustCreateSelfSignedCertSecret(t, "gateway-conformance-web-backend", "certificate", []string{"*"})
+	suite.Applier.MustApplyObjectsWithCleanup(t, suite.Client, suite.TimeoutConfig, []client.Object{secret}, suite.Cleanup)
+	secret = kubernetes.MustCreateSelfSignedCertSecret(t, "gateway-conformance-infra", "tls-validity-checks-certificate", []string{"*"})
+	suite.Applier.MustApplyObjectsWithCleanup(t, suite.Client, suite.TimeoutConfig, []client.Object{secret}, suite.Cleanup)
+	secret = kubernetes.MustCreateSelfSignedCertSecret(t, "gateway-conformance-infra", "tls-passthrough-checks-certificate", []string{"abc.example.com"})
+	suite.Applier.MustApplyObjectsWithCleanup(t, suite.Client, suite.TimeoutConfig, []client.Object{secret}, suite.Cleanup)
+
+	t.Logf("Test Setup: Ensuring Gateways and Pods from base manifests are ready")
+	namespaces := []string{
+		"gateway-conformance-infra",
+		"gateway-conformance-app-backend",
+		"gateway-conformance-web-backend",
+	}
+	kubernetes.NamespacesMustBeReady(t, suite.Client, suite.TimeoutConfig, namespaces)
+}
+
+// Run runs the provided set of conformance tests.
+func (suite *ExperimentalConformanceTestSuite) Run(t *testing.T, tests []ConformanceTest) error {
+	// verify that the test suite isn't already running, don't start a new run
+	// until the previous run finishes
+	suite.lock.Lock()
+	if suite.running {
```

Comment:
I may be missing something obvious, but what would lead to this being called multiple times? I'd assumed the primary way to run this would be via CLI. For reference, I'd assumed we'd have something similar to how Kubernetes conformance works where each implementation publishes the commands they ran to get the result they did: https://github.com/cncf/k8s-conformance/blob/master/v1.26/gke/README.md.

Reply:
The current suite simply does a `Run` and then it is done. This suite now has multiple knobs: you can run and then you can create a report from that run. Ultimately we don't _need_ to have this protection but for a low cost it makes the usage threadsafe in a future where there is more than one thing you would do with the object. I would suggest leaving this in unless there are some strong and specific objections because it makes the suite more robust.

Reply:
That makes sense, I think I'd just come into this review expecting something primarily CLI-based and this comment comes with that bias. This does seem reasonable if these tests are primarily intended to be run with code. I still think it would be helpful to have as simple as possible a way for others to run tests against an implementation to verify the results that have been reported. Admittedly that's not really related to the lines of code this comment was on.

Reply:
I hear you on the CLI thing. I've created #1900 for this purpose, which is now a sub-issue of #1709.